Add support for fine-grained retention policies (#1077)
tektoncd/results: docs/retention-policy-agent/README.md

# Result Retention Policy Agent

The Results Retention Policy Agent removes older Results and their associated Records from the DB. The policies apply to both `PipelineRun` and top-level `TaskRun` results.

## Configuration

The Results Retention Policy Agent is configured via the `tekton-results-config-results-retention-policy` ConfigMap.

The following fields are supported:

- `runAt`: Determines when to run the pruning job for the DB. It uses a cron schedule format. The default is `"7 7 * * 7"` (every Sunday at 7:07 AM).

> - `maxRetention`: The **fallback** retention period for how long to keep Results and Records when no specific policy matches. This value does **not** override the retention period of a matching policy; it only applies when no policies match a given Result. This can be a number (e.g., `30`), which is interpreted as days, or a duration string (e.g., `30d`, `24h`). The default is `30d`.

Author: True. But maxRetention won't work then because policy + defaultRetention will cover everything.

Maybe, I should deprecate maxRetention. 
 maxRetention in favour of defaultRetention name.

If maxRetention is given, assign it to defaultRetention.

> - `maxRetention`: The **fallback** retention period for how long to keep Results and Records when no specific policy matches. This value does **not** override the retention period of a matching policy; it only applies when no policies match a given Result. This can be a number (e.g., `30`), which is interpreted as days, or a duration string (e.g., `30d`, `24h`). The default is `30d`.

Author: Can you check the last commit?

The Results Retention Policy Agent removes older Results and their associated Records from the DB.

## Configuration

The Results Retention Policy Agent is configured via the `tekton-results-config-results-retention-policy` ConfigMap.

The following fields are supported:

- `runAt`: Determines when to run the pruning job for the DB. It uses a cron schedule format. The default is `"7 7 * * 7"` (every Sunday at 7:07 AM).

> - `maxRetention`: The default retention period for how long to keep Results and Records when no specific policy matches. This can be a number (e.g., `30`), which is interpreted as days, or a duration string (e.g., `30d`, `24h`). The default is `30d`.

Reviewer: `maxRetention` implies that it would also supersede any settings in `policies`. Is this correct? Do you mind clarifying how this behaves with specified retention policies?

Author: I will rewrite the doc to clarify that.

Author: I changed this to have it matched only Runs which aren't matched by any policy.  I can see that a feature request coming for that.  So let's do that instead now.

- `policies`: A list of fine-grained retention policies that allow for more specific control over data retention.

### Fine-Grained Retention Policies

> You can define a list of policies to control retention based on various criteria. The `policies` field in the ConfigMap accepts a YAML string containing a list of policy objects. Each policy has a `name`, a `selector`, and a `retention` period.

Reviewer: Does this need to be a string or could it be an object instead? Strings which are also yaml are poor UX in my opinion, especially if the string is already in a yaml document, since the yaml wont be validated it makes it difficult for CI to validate the settings are valid and easy for a user to introduce yaml errors.

Author: It's not a string TBH. It's marshal to a struct.

Author: We take a yaml and marshal it to a struct. Isn't that how we do with other configmap?

When the retention job runs, it evaluates a Result against the policies in the order they are defined. The **first policy that matches** the Result will be applied. If no policies match, the default `maxRetention` period is used.

#### Policy Fields:
- `name`: A descriptive name for the policy.
- `selector`: Defines the criteria for matching Results. All conditions within a selector must be met for the policy to match.

>   - `matchNamespace`: A list of namespaces. A Result matches if its namespace is in this list.

Reviewer: Nitpick: could this be `matchNamespaces` instead, since this is a list? The other fields are also plural

Author: Done,

>   - `matchLabels`: A map where the key is a label name and the value is a list of possible label values. A Result matches if it has a label with the given name and its value is in the list.

Author: It's AND  not OR.

>   - `matchLabels`: A map where the key is a label name and the value is a list of possible label values. A Result matches if it has a label with the given name and its value is in the list.

Author: `OR` is use for inside the labels. 
```
  matchLabels:
          "run": ["dev", "stage"]
          "test": ["yes"]
```

>   - `matchLabels`: A map where the key is a label name and the value is a list of possible label values. A Result matches if it has a label with the given name and its value is in the list.

Reviewer 2: What I meant is it's an OR inside the list.
That's very important clarification that should be added to the documentation, with the two examples you just added above.

>   - `matchLabels`: A map where the key is a label name and the value is a list of possible label values. A Result matches if it has a label with the given name and its value is in the list.

Author: I have updated the doc with this.

  - `matchAnnotations`: A map where the key is an annotation name and the value is a list of possible annotation values. A Result matches if it has an annotation with the given name and its value is in the list.

>   - `status`: A list of statuses (e.g., `Succeeded`, `Failed`, `Cancelled`). A Result matches if its final status is in this list.

Reviewer: Same request here for consistency. Could this be `statuses`? 

Also, are `Succeeded`, `Failed` and `Cancelled` a complete list of the valid statuses? I think it would be helpful to enumerate the allowed values here

Reviewer 2: Personally I like `status` more, as it resembles/matches the `status` field we have on the object.

Author: I changed to `matchStatuses`.

>   - `status`: A list of statuses (e.g., `Succeeded`, `Failed`, `Cancelled`). A Result matches if its final status is in this list.

Reviewer 2: I guess `Succeeded`, `Failed`, `Cancelled` is an ENUM, right, there aren't other statuses. If that's the case, I'll suggest:
 ```suggestion
  - `status`: A list of statuses (one or more of `Succeeded`, `Failed`, `Cancelled`). A Result matches if its final status is in this list.
```

Author: No, there are other statuses. Those are just common one.

Author: OK. I have updated the document to highlight this.

Author: I also added a link to the docs.
